lessons/intro/mypy/index.md

```diff
@@ -0,0 +1,357 @@
+# Mypy
+
+Dnes se společně podíváme na knihovnu `mypy`, která nám do Pythonu přidá
+možnost kontroly typů, jenž je jinak dostupná jen v jiných programovacích
+jazycích.
+
+## Statická a dynamická typová kontrola
+
+Statická typová kontrola v programovacím jazyce počítá s tím, že pro každou
+proměnnou či parametr je uveden datový typ. Proměnná definovaná s konkrétním
+typem pak nemůže obsahovat hodnotu jiného datového typu a pokus o to vyvolá
+výjimku. Jako např. tento kód v Javě:
+
+```java
+String name = "John";
+name = 5;
+```
+
+První řádek definuje proměnnou `name` jako řetězec. Na druhém řádku se
+pokoušíme do této proměnné přiřadit celé číslo, což není povoleno
+a kompilace takového kódu skončí chybou.
+
+Naproti tomu Python je typickým představitelem jazyků s dynamickou typovou
+kontrolou. Program jako je ten následující bude fungovat naprosto
+bez problémů:
+
+```python
+name = "John"
+name = 5
+name = ["John", "Peter", "Alice"]
+```
+
+Python obecně spoléhá na to, že si sám tvůrce dá pozor, aby proměnná
+obsahovala v tom správném čase ten správný obsah a v případě nesrovnalostí
+se s tím program za běhu nějak popasuje.
+
+Oba přístupy mají své výhody a nevýhody. Statická typová kontrola by možná
+pro začátečníky ubrala na čitelnosti kódu, zbytečně prodloužila krátké
+jednoúčelové programy a nejspíše by ani nenadchla vědce či matematiky, kteří
+v Pythonu tvoří většinu času prototypy, které nemají testy.
+Na druhou stranu může přispět k lepší čitelnosti složitějšího kódu,
+editory a IDE mohou díky ní lépe napovídat programátorům a testování
```

Comment:
`testování`? Jak je to myšleno?

Reply:
Myšleno testování definovaných typů, při kterém se třeba odhalí, že funkce vrací None a ty tuto hodnotu dáváš další funkci, která potřebuje na vstupu string.

Reply:
To bych nazval spíš automatickou kontrolou typů, nebo možná validací. Pod testováním si představím, že se něco fakt testuje :) tj. že se to nějak aktivně zkouší (zde např. přímo spouštěním kódu).

Reply:
Trošku jsem ten odstavec přepsal, aby z toho vyplynulo, že typová kontrola je samostatná operace.